ports/esp32/network_ppp.c

```diff
@@ -0,0 +1,234 @@
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+
+#include "py/runtime.h"
+#include "py/mphal.h"
+#include "py/objtype.h"
+#include "netutils.h"
+
+#include "modmachine.h"
+#include "machine_uart.h"
+
+#include "strings.h"
+
+#include "netif/ppp/ppp.h"
+#include "netif/ppp/pppos.h"
+#include "lwip/err.h"
+#include "lwip/sockets.h"
+#include "lwip/sys.h"
+#include "lwip/netdb.h"
+#include "lwip/dns.h"
+#include "lwip/pppapi.h"
+
+extern const mp_obj_type_t machine_uart_type;
+
+typedef struct _ppp_if_obj_t {
+    mp_obj_base_t base;
+    bool initialized;
+    bool active;
+    bool connected;
+    ppp_pcb *pcb;
+    machine_uart_obj_t* uart;
+    struct netif pppif;
+    SemaphoreHandle_t inactiveWaitSem;
+    TaskHandle_t client_task_handle;
+} ppp_if_obj_t;
+
+const mp_obj_type_t ppp_if_type;
+STATIC ppp_if_obj_t ppp_obj = {{&ppp_if_type}, false, false, false, NULL, NULL};
+
+/* PPP status callback example */
+static void ppp_status_cb(ppp_pcb *pcb, int err_code, void *ctx) // wait for
+{
+    ppp_if_obj_t* self = &ppp_obj;
+    struct netif *pppif = ppp_netif(self->pcb);
+    (void)pppif;
+    LWIP_UNUSED_ARG(ctx);
+
+    switch (err_code) {
+        case PPPERR_NONE: {
+                self->connected = (pppif->ip_addr.u_addr.ip4.addr != 0);
+                break;
+            }
+        case PPPERR_USER: {
+                xSemaphoreGive(self->inactiveWaitSem);
+                break;
+            }
+        case PPPERR_CONNECT: {
+                self->connected = false;
+                break;
+            }
+        default: {
+                break;
+            }
+    }
+}
+
+STATIC mp_obj_t get_ppp(mp_obj_t uart) {
+    ppp_if_obj_t* self = &ppp_obj;
+
+    if (self->initialized) {
+        return MP_OBJ_FROM_PTR(&ppp_obj);
+    }
+
+    if (! MP_OBJ_IS_TYPE(uart, &machine_uart_type)) {
+        mp_raise_TypeError("requires machine.UART type");
+    }
+
+    self->uart = (machine_uart_obj_t*) uart;
+    self->inactiveWaitSem = xSemaphoreCreateBinary();
+    assert(self->inactiveWaitSem != NULL);
+    self->initialized = true;
+    return MP_OBJ_FROM_PTR(&ppp_obj);
+}
+MP_DEFINE_CONST_FUN_OBJ_1(get_ppp_obj, get_ppp);
+
+static u32_t ppp_output_callback(ppp_pcb *pcb, u8_t *data, u32_t len, void *ctx)
+{
+    ppp_if_obj_t* self = &ppp_obj;
+    mp_stream_p_t* uart_stream = (mp_stream_p_t*)self->uart->base.type->protocol;
+    int err;
+
+    return uart_stream->write(self->uart, data, len, &err);
+}
+
+static void pppos_client_task() {
+    ppp_if_obj_t* self = &ppp_obj;
+    self->uart->timeout = 50;
+    mp_stream_p_t* uart_stream = (mp_stream_p_t*)self->uart->base.type->protocol;
+    uint8_t buf[256];
+
+    int err;
+    while (ulTaskNotifyTake(pdTRUE, 0) == 0) {
```

Comment:
I assume this is used to signal the task to exit?

Reply:
Yup, FreeRTOS has some sort of built-in per task semaphore, didn't see any reason not to use it in this case.